Libplanet/Blockchain/BlockChain.cs

```diff
@@ -465,21 +481,57 @@ DateTimeOffset currentTime
                 .Select(Store.GetTransaction<T>)
                 .Where(tx => tx.Nonce < GetNextTxNonce(tx.Signer));
 
-            Block<T> block = Block<T>.Mine(
-                index: index,
-                difficulty: difficulty,
-                miner: miner,
-                previousHash: prevHash,
-                timestamp: currentTime,
-                transactions: transactions
-            );
+            CancellationTokenSource cts = new CancellationTokenSource();
+            CancellationTokenSource cancellationTokenSource =
+                CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cts.Token);
+
+            void WatchTip(object target, long tipIndex)
+            {
+                cts.Cancel();
+            }
+
+            _tipChanged += WatchTip;
+
+            Block<T> block;
+            try
+            {
+                Task<Block<T>> t = Task.Run(
+                    () => Block<T>.Mine(
+                        index: index,
+                        difficulty: difficulty,
+                        miner: miner,
+                        previousHash: prevHash,
+                        timestamp: currentTime,
+                        transactions: transactions,
+                        cancellationToken: cancellationTokenSource.Token),
+                    cancellationTokenSource.Token
+                );
+                block = await t;
+            }
+            catch (OperationCanceledException)
+            {
+                if (cts.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException(
+                        "Mining canceled due to change of tip index");
+                }
+                else
```

Comment:
How about just re-throw it?

```suggestion
                throw;
```

Reply:
I used `TaskCanceledException` here because other tasks used in `Swarm<T>` throw `TaskCanceledException` when cancel is requested.